src/resolver.rs

```diff
@@ -0,0 +1,44 @@
+//! Abstractions over DNS resolvers.
+
+use std::{
+    future::Future,
+    net::{SocketAddr, ToSocketAddrs},
+    pin::Pin,
+};
+
+use crate::Error;
+
+/// Trait for a DNS resolver to resolve hostnames and ports to IP addresses.
+pub trait Resolver {
+    /// Resolve a hostname and port to an IP address, asynchronously.
+    fn resolve(
+        &self,
+        host: &str,
+        port: u16,
+    ) -> Pin<Box<dyn Future<Output = Result<SocketAddr, Error>>>>;
+}
+
+/// A [`Resolver`] that uses the blocking `getaddrinfo` syscall in the tokio
+/// threadpool.
+pub struct Gai;
+
+impl Resolver for Gai {
+    fn resolve(
+        &self,
+        host: &str,
+        port: u16,
+    ) -> Pin<Box<dyn Future<Output = Result<SocketAddr, Error>>>> {
+        let host = host.to_owned();
+
+        Box::pin(async move {
+            let task = tokio::task::spawn_blocking(move || {
+                (host, port)
+                    .to_socket_addrs()?
+                    .next()
+                    .ok_or(Error::CannotResolveHost)
```

Comment:
```suggestion
            let mut addrs = tokio::net::lookup_host((host, port)).await?;

            addrs.next().ok_or(Error::CannotResolveHost)
```

Is there a reason that you don't use `lookup_host`?